### ISSUE TITLE
[Question] Parallel Learning PPO

### ISSUE THREAD

**Issue by danielstankw** [DLR-RM/stable-baselines3#767]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

### Question

I am using a robosuite which is a program for RL that is based on mujoco simulation engine. I am trying to run a parallel learning using 5 cpu's but for some reason it doesnt work. 
I can see that the simulation does run 5 processes at the time, but the total learning time is exactly the same as when I train it using 1 process. I know that I shouldnt expect the time to be 5 times shorter, but at least by few hours, but in my case it doesnt matter whether I run it with 1 cpu or 5, it takes same amount of time.

Any clues what can be the issue? 

```
def make_robosuite_env(env_id, options, rank, seed=0):
    """
    Utility function for multiprocessed env.

    :param env_id: (str) the environment ID
    :param options: additional arguments to pass to the specific env class initializer
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    """

    def _init():
        env = GymWrapper(suite.make(env_id, **options))
        env = Monitor(env)
        # Important: use a different seed for each environment
        env.seed(seed + rank)
        return env

    set_random_seed(seed)
    return _init

control_param = dict(type='IMPEDANCE_POSE_Partial', input_max=1, input_min=-1,
                         output_max=[0.05, 0.05, 0.05, 0.5, 0.5, 0.5],
                         output_min=[-0.05, -0.05, -0.05, -0.5, -0.5, -0.5], kp=700, damping_ratio=np.sqrt(2),
                         impedance_mode='fixed', kp_limits=[0, 100000], damping_ratio_limits=[0, 10],
                         position_limits=None, orientation_limits=None, uncouple_pos_ori=True, control_delta=True,
                         interpolation=None, ramp_ratio=0.2, control_dim=dim_c, plotter=False, ori_method='rotation')

env_options = dict(robots="UR5e", use_camera_obs=False, has_offscreen_renderer=False, has_renderer=render,
                         reward_shaping=True, ignore_done=False, horizon=horizon, control_freq=20,
                         controller_configs=control_param, r_reach_value=0.2, tanh_value=20.0, error_type=error_type,
                         control_spec=dim_c, dist_error=err)

# creating multiprocessing environments
num_cpu = 5  # max: 8
env = SubprocVecEnv([make_robosuite_env(env_id, env_options, i, my_seed) for i in range(num_cpu)])
n_steps = 10

policy_kwargs = dict(activation_fn=torch.nn.LeakyReLU, net_arch=[32, 32])

model = PPO(MlpPolicy, env, verbose=1, n_steps=int(n_steps/num_cpu), batch_size=10, seed=my_seed, policy_kwargs=policy_kwargs, tensorboard_log="./learning_log/ppo_tensorboard/")

print('Multi-environment learning')
model.learn(total_timesteps=learning_timesteps, tb_log_name="learning", reset_num_timesteps=True)
```

### Additional context

Add any other context about the question here.

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by danielstankw** [comment]
![image](https://user-images.githubusercontent.com/72759092/153560953-8152cbef-4028-4927-85fe-24df3ef7cb0e.png)

> I think this is mentioned somewhere in the docs, but SubProcVecEnv does not provide speed boost if the environment is fast to run (which MuJoCo is). The overhead of inter-process communication outweighs the benefit. Try using DummyVecEnv instead.

I use the SubProcvecEnv as I saw someone who also worked with the mujoco used it and they succeeded in faster learning. Also according to the docs the DummyVecEnv doesn't support multiprocessing up to my understanding. 

> Also set the algorithm to use cpu, because with small envs GPU bottleneck is huge.

I didn't set it explicitly to use GPU so by default it should use CPU. You know how can I check if its using CPU or GPU? 

> It is also possible that your training is the bottleneck: your batch size is very small, which definitely slows things down.

Good point, I will try to play around with this value, and see if changing it will help

**Comment by Miffyli** [comment]
> I use the SubProcvecEnv as I saw someone who also worked with the mujoco used it and they succeeded in faster learning. Also according to the docs the DummyVecEnv doesn't support multiprocessing up to my understanding.

Yup you are right. However DummyVecEnv can still yield bigger speed boost than SubProcVecEnv because a) it does not have the multi-process communication overhead and b) using bigger batches of data (which you get with more envs) makes better use of hardware during training (i.e. processes more samples per time unit when batches are bigger).

> I didn't set it explicitly to use GPU so by default it should use CPU. You know how can I check if its using CPU or GPU?

By default GPU is used if CUDA-enabled PyTorch is installed. You can check the `agent.device` attribute for this (or, use `nvidia-smi` to see if GPU is being used)

**Comment by gunnxx** [comment]
Hi @Miffyli , I am wondering the details of your state,emt `Also set the algorithm to use cpu, because with small envs GPU bottleneck is huge`. If I am using off-policy algorithms and do more gradient updates per step in the environment, is it still good to use CPU?

**Comment by Miffyli** [comment]
@gunnxx I can not say for sure but I would not be surprised if CPU works faster even with off-policy algorithms. I suggest you give it a try :).

And yup, at least MuJoCo and other common physics engine envs do seem to work fine with `DummyVecEnv`.
